### PR TITLE
Update to support latest vsomeip and Zenoh transports

### PR DESCRIPTION
**Pull request by PLeVasseur**
~_Note_: Opening as draft pull request as we need to have https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust/pull/31 merge first.~

Above PR merged, so removed draft status.

### Testing Status of Examples

I did a manual test of the examples to make sure that communications were flowing.

|                               | ✅ or ❌ |
|-------------------------------|--------|
| me_client <-> ue_service      | ✅      |
| ue_client <-> me_service      | ✅      |
| me_publisher -> ue_subscriber | ✅      |
| ue_publisher -> me_subscriber | ✅      |

### CODE REVIEW

File: example-streamer-uses/Cargo.toml [hunk 1]
```diff
@@ -55,6 +55,6 @@ protobuf = { workspace = true }
 serde = { workspace = true }
 tokio = { workspace = true }
 up-rust = { workspace = true }
-up-transport-zenoh = { version = "0.1.1"  }
-up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.1.0", default-features = false }
-zenoh = { version = "1.0.0-alpha.6", features = ["unstable"]}
+up-transport-zenoh = { version = "0.3.0"  }
+up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.3.0", default-features = false }
+zenoh = { version = "1.0.0", features = ["unstable", "internal"] }
```
Comment (sophokles73) [inline example-streamer-uses/Cargo.toml:60]: FMPOV we should no longer depend on _unstable_ features and internal API. This only makes migration/updating to newer Zenoh versions harder ...

File: up-linux-streamer-plugin/Cargo.toml [hunk 2]
```diff
@@ -44,14 +44,14 @@ serde = { version = "1.0.154" }
...
+up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.3.0", default-features = false }
 up-streamer = { path = "../up-streamer" }
 usubscription-static-file = {path = "../utils/usubscription-static-file"}
-zenoh = { version = "1.0.0-alpha.6", features = ["unstable", "internal", "plugins"]}
-zenoh-core = { version = "1.0.0-alpha.6" }
-zenoh-plugin-trait = { version = "1.0.0-alpha.6" }
-zenoh-result = { version = "1.0.0-alpha.6" }
-zenoh-util = { version = "1.0.0-alpha.6" }
-zenoh_backend_traits = { version = "1.0.0-alpha.6" }
+zenoh = { version = "1.0.0", features = ["unstable", "internal", "plugins"] }
```
Comment (sophokles73) [inline up-linux-streamer-plugin/Cargo.toml:51]: see previous comment

File: up-linux-streamer/Cargo.toml [hunk 3]
```diff
@@ -35,9 +35,9 @@ serde = { workspace = true }
 tokio = { workspace = true }
 up-rust = { workspace = true }
 up-streamer = { path = "../up-streamer" }
-up-transport-zenoh = { version = "0.1.1"  }
-up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.1.0", default-features = false }
-zenoh = { version = "1.0.0-alpha.6", features = ["unstable"]}
+up-transport-zenoh = { version = "0.3.0"  }
+up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.3.0", default-features = false }
+zenoh = { version = "1.0.0", features = ["unstable", "internal"] }
```
Comment (sophokles73) [inline up-linux-streamer/Cargo.toml:40]: and again

File: up-streamer/src/ustreamer.rs [hunk 4]
```diff
@@ -34,19 +38,20 @@ const USTREAMER_FN_NEW_TAG: &str = "new():";
...
-        ue_version_major: 0xFF, // any
-        resource_id: 0xFFFF,    // any
-        ..Default::default()
-    }
+const THREAD_NUM: usize = 10;
+
+// Create a separate tokio Runtime for running the callback
+lazy_static! {
+    static ref CB_RUNTIME: Runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(THREAD_NUM)
```
Comment (sophokles73) [inline up-streamer/src/ustreamer.rs:46]: isn't there a reasonable default in Tokio, e.g. based on the number of cores?

File: example-streamer-uses/Cargo.toml (hunk 1 above)
Comment (PLeVasseur) [reply example-streamer-uses/Cargo.toml:60]: Let me investigate if we need these still. We used to need.

File: up-linux-streamer-plugin/Cargo.toml (hunk 2 above)
Comment (PLeVasseur) [reply up-linux-streamer-plugin/Cargo.toml:51]: The plugins one I think we need.

But I'm thinking to generally only compile the reference implementations of the streamers conditionally going forward.

Will help with needing to compile vsomeip as well.

File: up-streamer/src/ustreamer.rs (hunk 4 above)
Comment (PLeVasseur) [reply up-streamer/src/ustreamer.rs:46]: I'd like to make this configurable as I did in the vsomeip transport. Opened #56 to track this, will mark as resolved for now.

File: example-streamer-uses/Cargo.toml (hunk 1 above)
Comment (PLeVasseur) [reply example-streamer-uses/Cargo.toml]: Okay! I was able to remove both "unstable" and "internal".

File: up-linux-streamer-plugin/Cargo.toml (hunk 2 above)
Comment (PLeVasseur) [reply up-linux-streamer-plugin/Cargo.toml]: Needed to keep the "plugins" feature, but again, was able to remove "internal" and "unstable".

File: up-linux-streamer/Cargo.toml (hunk 3 above)
Comment (PLeVasseur) [reply up-linux-streamer/Cargo.toml]: Was able to remove both "unstable" and "internal".

File: up-linux-streamer-plugin/Cargo.toml (hunk 2 above)
Comment (PLeVasseur) [reply up-linux-streamer-plugin/Cargo.toml]: Updated to compile the reference implementations of the streamers in full conditionally based on a feature flag.

File: up-linux-streamer-plugin/Cargo.toml [hunk 5]
```diff
@@ -48,7 +48,7 @@ up-transport-zenoh = { version = "0.3.0"  }
 up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.3.0", default-features = false }
 up-streamer = { path = "../up-streamer" }
 usubscription-static-file = {path = "../utils/usubscription-static-file"}
-zenoh = { version = "1.0.0", features = ["unstable", "internal", "plugins"] }
+zenoh = { version = "1.0.0", features = ["default", "plugins"] }
```
Comment (sophokles73) [inline up-linux-streamer-plugin/Cargo.toml:53]: how about defining the Zenoh dependency at the workspace level?

File: .github/workflows/bundled-lint-and-test.yaml [hunk 6]
```diff
@@ -63,9 +63,12 @@ jobs:
           rustup show
           rustup component add rustfmt clippy
 
-      - name: Build the project
+      - name: Build the project without Zenoh & vsomeip streamer reference
```
Comment (sophokles73) [inline .github/workflows/bundled-lint-and-test.yaml:66]: wouldn't that rather be
> without Zenoh & vsomeip transport reference

Actually, it seems like you only moved the dependency to the vsomeip trnasport behind a feature ...

File: up-linux-streamer-plugin/Cargo.toml (hunk 5 above)
Comment (PLeVasseur) [reply up-linux-streamer-plugin/Cargo.toml:53]: Yeah, after doing this yesterday and your comment, I think that's what I'll try to do. I'm going to try to define some of the transport specific stuff at the workspace-level (including Zenoh), make them optional and then again optionally include them based on the same feature flag. Let's see how that goes.

File: .github/workflows/bundled-lint-and-test.yaml (hunk 6 above)
Comment (PLeVasseur) [reply .github/workflows/bundled-lint-and-test.yaml:66]: Yeah, I'm reconsidering how to do this. Maybe one feature flag per transport. And then if both are enabled we build the streamer references as well.

File: up-linux-streamer-plugin/Cargo.toml (hunk 5 above)
Comment (PLeVasseur) [reply up-linux-streamer-plugin/Cargo.toml:53]: Hmm, seems like it's not possible to define optional dependencies at the workspace level. I'm looking to go the other direction. Lemme try some stuff.

Comment (PLeVasseur) [reply up-linux-streamer-plugin/Cargo.toml]: Could you take another look at where I ended up?

File: .github/workflows/bundled-lint-and-test.yaml (hunk 6 above)
Comment (PLeVasseur) [reply .github/workflows/bundled-lint-and-test.yaml]: Resolved this comment, looking for more feedback!

File: .github/workflows/unbundled-lint-and-test.yaml [hunk 7]
```diff
@@ -160,15 +160,21 @@ jobs:
...
+        run: cargo build
+      - name: cargo clippy without Zenoh & vsomeip transports
+        working-directory: ${{github.workspace}}
+        run: cargo clippy --all-targets -- -W warnings -D warnings
+      - name: Build the project with Zenoh & vsomeip transports (and thus streamer references)
+        working-directory: ${{github.workspace}}
+        run: cargo build --features enable-vsomeip-transport,bundled-vsomeip,enable-zenoh-transport
+      - name: cargo clippy with Zenoh & vsomeip transports (and thus streamer references)
+        working-directory: ${{github.workspace}}
+        run: cargo clippy --features enable-vsomeip-transport,bundled-vsomeip,enable-zenoh-transport --all-targets -- -W warnings -D warnings
```
Comment (sophokles73) [inline .github/workflows/unbundled-lint-and-test.yaml:174]: Is there any difference anymore between this and `bundled-lint-and-test.yaml`?

File: Cargo.toml [hunk 8]
```diff
@@ -33,14 +33,15 @@ async-trait = { version = "0.1" }
...
 env_logger = { version = "0.10.1" }
 futures = { version = "0.3.30" }
+lazy_static = { version = "1.5.0" }
 log = { version = "0.4.20" }
 json5 = { version =  "0.4.1" }
 serde = { version = "1.0.154", features = ["derive"] }
 serde_json = { version = "1.0.94" }
 uuid = { version = "1.7.0" }
-tokio = { version = "1.35.1", default-features = false }
+tokio = { version = "1.35.1", default-features = false, features = ["rt", "rt-multi-thread", "sync", "time"] }
```
Comment (sophokles73) [inline Cargo.toml:42]: As a general pattern, I tend to declare at the workspace level with `default-features = false` and then enable features at the package level. That way everybody just gets what they need ...

File: example-streamer-uses/Cargo.toml [hunk 9]
```diff
@@ -55,6 +68,6 @@ protobuf = { workspace = true }
 serde = { workspace = true }
 tokio = { workspace = true }
 up-rust = { workspace = true }
-up-transport-zenoh = { version = "0.1.1"  }
-up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.1.0", default-features = false }
-zenoh = { version = "1.0.0-alpha.6", features = ["unstable"]}
+up-transport-zenoh = { version = "0.3.0", optional = true }
+up-transport-vsomeip = { git = "https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust.git", tag = "v0.3.0", default-features = false, optional = true }
+zenoh = { version = "1.0.0", optional = true }
```
Comment (sophokles73) [inline example-streamer-uses/Cargo.toml:73]: Any particular reason why you do not want to declare these at the workspace level and then add the `optional = true` flag at the package level ...

File: example-streamer-uses/Cargo.toml [hunk 10]
```diff
@@ -21,27 +21,40 @@ license.workspace = true
...
 name = "ue_service"
+required-features = ["enable-zenoh-transport"]
 
 [[bin]]
 name = "ue_subscriber"
+required-features = ["enable-zenoh-transport"]
+
+[features]
+enable-zenoh-transport = ["up-transport-zenoh", "zenoh"]
+enable-vsomeip-transport = ["up-transport-vsomeip"]
```
Comment (sophokles73) [inline example-streamer-uses/Cargo.toml:56]: How about just calling these `zenoh-transport` and `vsomeip-transport`? A feature needs to be enabled (by supplying the `cargo --features` parameter), but the feature itself should just be named according to the functionality it provides FMPOV ...

File: Cargo.toml (hunk 8 above)
Comment (PLeVasseur) [reply Cargo.toml:42]: Yup, I agree that would be the way to go normally. I'll outline here why I didn't go that route.

* I want to allow us to compile the workspace project without relying on bringing in the vsomeip or Zenoh at all, if someone is interested in only working with / on the `up-streamer` generic library.
  * In particular... the vsomeip one. It's annoying at the moment to compile due to requiring us to supply the locations of the C++ headers of the machine. This is something I should look into fixing: opened issue here: https://github.com/eclipse-uprotocol/up-transport-vsomeip-rust/issues/33.
* A Cargo workspace does not allow us to pick and choose members to compile based on feature flags.
* A Cargo workspace does not allow us to optionally include dependencies based on feature flags.

This left me with the not great solution of pushing these dependencies down into each crate and use the same feature flags across the three crates: `example-streamer-uses`, `up-linux-streamer`, `up-linux-streamer-plugin`.

When the up-transport-vsomeip compilation story is less bad, I'd like to do as you suggest.

File: example-streamer-uses/Cargo.toml (hunk 9 above)
Comment (PLeVasseur) [reply example-streamer-uses/Cargo.toml:73]: Wrote that up [here](https://github.com/eclipse-uprotocol/up-streamer-rust/pull/54#discussion_r1842180856).

File: example-streamer-uses/Cargo.toml (hunk 10 above)
Comment (PLeVasseur) [reply example-streamer-uses/Cargo.toml:56]: Okay by me.

File: .github/workflows/unbundled-lint-and-test.yaml (hunk 7 above)
Comment (PLeVasseur) [reply .github/workflows/unbundled-lint-and-test.yaml:174]: I... think you're right. I should not have included the `bundled-vsomeip` flag here!

**Review by sophokles73 — APPROVED** [review]
LGTM

don't forget to squash ...

File: .github/workflows/unbundled-lint-and-test.yaml (hunk 7 above)
Comment (PLeVasseur) [reply .github/workflows/unbundled-lint-and-test.yaml]: Done

**Pull request merged**